compiler/src/iree/compiler/DispatchCreation/FoldReshapesIntoTensorBarriers.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/TensorExt/IR/TensorExtOps.h"
+#include "iree/compiler/DispatchCreation/Passes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::DispatchCreation {
+
+#define GEN_PASS_DEF_FOLDRESHAPESINTOTENSORBARRIERSPASS
+#include "iree/compiler/DispatchCreation/Passes.h.inc"
+
+namespace {
+
+// Move tensor.expand_shape/collapse_shape above compute_barrier.start
+struct MoveReshapeAboveBarrierStart : public RewritePattern {
+  MoveReshapeAboveBarrierStart(MLIRContext *context, PatternBenefit benefit = 1)
+      : RewritePattern(MatchAnyOpTypeTag(), benefit, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!isa<tensor::ExpandShapeOp, tensor::CollapseShapeOp>(op)) {
+      return failure();
+    }
+
+    auto barrierStartOp =
+        op->getOperand(0)
+            .getDefiningOp<IREE::TensorExt::ComputeBarrierStartOp>();
+    if (!barrierStartOp) {
+      return failure();
+    }
+
+    // Update reshape's operand to use compute_barrier.start's input
+    rewriter.modifyOpInPlace(
+        op, [&]() { op->getOpOperand(0).set(barrierStartOp.getValue()); });
+
+    IRRewriter::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointAfter(op);
+    Value reshapeResult = op->getResult(0);
+    auto newBarrier = IREE::TensorExt::ComputeBarrierStartOp::create(
+        rewriter, barrierStartOp.getLoc(), reshapeResult);
+    reshapeResult.replaceUsesWithIf(
```

Comment:
I think its better to go through the rewriter to do this replace. It sometimes makes the pattern rewriter behave strangely when you dont do that.